modules/host/negotiatedownload.go

```diff
@@ -0,0 +1,230 @@
+	if err != nil {
+		return err
+	}
+
+	// Verify that the request is acceptable, and then fetch all of the data
+	// for the renter.
+	var payload [][]byte
+	var hostSignature types.TransactionSignature
+	err = func() error {
+		// Check that the length of each file is in-bounds, and that the total
+		// size being requested is acceptable.
+		var totalSize uint64
+		for _, request := range requests {
+			if request.Length > modules.SectorSize || request.Offset+request.Length > modules.SectorSize {
+				return errRequestOutOfBounds
+			}
+			totalSize += request.Length
+		}
+		if totalSize > settings.MaxDownloadBatchSize {
+			return errLargeDownloadBatch
+		}
+
+		// Verify that the correct amount of money has been moved from the
+		// renter's contract funds to the host's contract funds.
+		expectedTransfer := settings.MinimumDownloadBandwidthPrice.Mul(types.NewCurrency64(totalSize))
+		existingRevision := so.RevisionTransactionSet[len(so.RevisionTransactionSet)-1].FileContractRevisions[0]
+		// Check that NewValidProofOutputs[0] is deducted by expectedTransfer.
+		if paymentRevision.NewValidProofOutputs[0].Value.Add(expectedTransfer).Cmp(existingRevision.NewValidProofOutputs[0].Value) != 0 {
+			return errors.New("bad payments")
+		}
+		// Check that NewValidProofOutputs[1] is incremented by expectedTransfer.
+		if existingRevision.NewValidProofOutputs[1].Value.Add(expectedTransfer).Cmp(paymentRevision.NewValidProofOutputs[1].Value) != 0 {
+			return errors.New("bad payouts")
+		}
+		// Check that MissedProofOutputs[0] is decremented by expectedTransfer.
+		if paymentRevision.NewMissedProofOutputs[0].Value.Add(expectedTransfer).Cmp(existingRevision.NewMissedProofOutputs[0].Value) != 0 {
+			return errors.New("bad payments")
+		}
+		// Check that NewMissedProofOutputs[2] is incremented by expectedTransfer.
+		if existingRevision.NewMissedProofOutputs[2].Value.Add(expectedTransfer).Cmp(paymentRevision.NewMissedProofOutputs[2].Value) != 0 {
+			return errors.New("bad payouts")
+		}
+		// Check that the revision count has increased.
+		if paymentRevision.NewRevisionNumber <= existingRevision.NewRevisionNumber {
+			return errors.New("bad revision number")
+		}
+
+		// Check that no other fields have changed.
+		if paymentRevision.ParentID != existingRevision.ParentID {
+			return errors.New("unstable revision")
+		}
+		if paymentRevision.NewFileSize != existingRevision.NewFileSize {
+			return errors.New("unstable revision")
+		}
+		if paymentRevision.NewFileMerkleRoot != existingRevision.NewFileMerkleRoot {
+			return errors.New("unstable revision")
+		}
+		if paymentRevision.NewWindowStart != existingRevision.NewWindowStart {
+			return errors.New("unstable revision")
+		}
+		if paymentRevision.NewWindowEnd != existingRevision.NewWindowEnd {
+			return errors.New("unstable revision")
+		}
+		if paymentRevision.NewUnlockHash != existingRevision.NewUnlockHash {
+			return errors.New("unstable revision")
```

Comment:
it's not done, they were thrown there quickly so you would have a reference to write your code against. Fixing up in more detail now.